File: src/main/java/ru/citeck/ecos/uiserv/domain/journal/api/records/JournalRecordsDao.java
package ru.citeck.ecos.uiserv.domain.journal.api.records;

import ecos.com.fasterxml.jackson210.annotation.JsonIgnore;
import ecos.com.fasterxml.jackson210.annotation.JsonProperty;
import ecos.com.fasterxml.jackson210.annotation.JsonValue;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import org.apache.commons.lang.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.springframework.stereotype.Component;
import ru.citeck.ecos.commons.data.MLText;
import ru.citeck.ecos.commons.data.ObjectData;
import ru.citeck.ecos.commons.json.Json;
import ru.citeck.ecos.records2.QueryContext;
import ru.citeck.ecos.records2.RecordMeta;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records2.graphql.meta.annotation.MetaAtt;
import ru.citeck.ecos.records2.graphql.meta.value.MetaField;
import ru.citeck.ecos.records2.predicate.PredicateService;
import ru.citeck.ecos.records2.predicate.model.Predicate;
import ru.citeck.ecos.records2.request.delete.RecordsDelResult;
import ru.citeck.ecos.records2.request.delete.RecordsDeletion;
import ru.citeck.ecos.records2.request.mutation.RecordsMutResult;
import ru.citeck.ecos.records2.request.query.RecordsQuery;
import ru.citeck.ecos.records2.request.query.RecordsQueryResult;
import ru.citeck.ecos.records2.source.dao.local.LocalRecordsDao;
import ru.citeck.ecos.records2.source.dao.local.MutableRecordsLocalDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsMetaDao;
import ru.citeck.ecos.records2.source.dao.local.v2.LocalRecordsQueryWithMetaDao;
import ru.citeck.ecos.uiserv.domain.journal.dto.ColumnControl;
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalColumnDto;
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalDto;
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalWithMeta;
import ru.citeck.ecos.uiserv.domain.journal.service.JournalService;
import ru.citeck.ecos.uiserv.domain.journal.service.type.TypeJournalService;

import java.util.*;
import java.util.stream.Collectors;

@Component
@RequiredArgsConstructor
public class JournalRecordsDao extends LocalRecordsDao
                               implements LocalRecordsQueryWithMetaDao<JournalWithMeta>,
                                          LocalRecordsMetaDao<JournalWithMeta>,
                                          MutableRecordsLocalDao<JournalRecordsDao.JournalRecord> {

    public static final String LANG_QUERY_BY_LIST_ID = "list-id";
    public static final String LANG_QUERY_SITE_JOURNALS = "site-journals";

    private final JournalService journalService;
    private final TypeJournalService typeJournalService;

    @Override
    public RecordsQueryResult<JournalWithMeta> queryLocalRecords(@NotNull RecordsQuery recordsQuery,
                                                                 @NotNull MetaField metaField) {

        RecordsQueryResult<JournalWithMeta> result = new RecordsQueryResult<>();

        if (LANG_QUERY_BY_LIST_ID.equals(recordsQuery.getLanguage())) {

            QueryByListId queryByListId = recordsQuery.getQuery(QueryByListId.class);
            List<JournalWithMeta> journals = journalService.getJournalsByJournalList(queryByListId.getListId());
            journals.forEach(j -> result.addRecord(new JournalRecord(j)));

            return result;

        } else if (LANG_QUERY_SITE_JOURNALS.equals(recordsQuery.getLanguage())) {

            journalService.getJournalsWithSite().forEach(j -> result.addRecord(new JournalRecord(j)));
            return result;
        }

        JournalQueryByTypeRef queryByTypeRef = recordsQuery.getQuery(JournalQueryByTypeRef.class);
        if (queryByTypeRef != null && queryByTypeRef.getTypeRef() != null) {
            typeJournalService.getJournalForType(queryByTypeRef.getTypeRef()).ifPresent(dto ->
                result.addRecord(new JournalRecord(dto))
            );
            return result;
        }

        if (recordsQuery.getLanguage().equals(PredicateService.LANGUAGE_PREDICATE)) {
            Predicate predicate = recordsQuery.getQuery(Predicate.class);

            int max = recordsQuery.getMaxItems();
            if (max <= 0) {
                max = 10000;
            }

            List<JournalWithMeta> journals = journalService.getAll(max, recordsQuery.getSkipCount(), predicate);

            result.setRecords(new ArrayList<>(journals));
            result.setTotalCount(journalService.getCount(predicate));

        } else {
            result.setRecords(new ArrayList<>(
                journalService.getAll(recordsQuery.getMaxItems(), recordsQuery.getSkipCount()))
            );
            result.setTotalCount(journalService.getCount());
        }

        return new RecordsQueryResult<>(result, JournalRecord::new);
    }

    @Override
    public List<JournalWithMeta> getLocalRecordsMeta(@NotNull List<RecordRef> list,
                                                     @NotNull MetaField metaField) {

        return list.stream()
            .map(ref -> {
                JournalWithMeta dto;
                if (RecordRef.isEmpty(ref)) {
                    dto = new JournalWithMeta();
                } else {
                    String id = ref.getId();
                    if (id.startsWith(TypeJournalService.JOURNAL_ID_PREFIX)) {
                        RecordRef typeRef = RecordRef.create(
                            "emodel",
                            "type",
                            id.substring(TypeJournalService.JOURNAL_ID_PREFIX.length())
                        );
                        dto = typeJournalService.getJournalForType(typeRef).orElse(null);
                    } else {
                        dto = journalService.getById(ref.getId());
                    }
                    if (dto == null) {
                        dto = new JournalWithMeta();
                    }
                }
                return new JournalRecord(dto);
            }).collect(Collectors.toList());
    }

    @Override
    public RecordsDelResult delete(RecordsDeletion deletion) {

        List<RecordMeta> resultRecords = new ArrayList<>();

        deletion.getRecords()
            .forEach(r -> {
                journalService.delete(r.getId());
                resultRecords.add(new RecordMeta(r));
            });

        RecordsDelResult result = new RecordsDelResult();
        result.setRecords(resultRecords);
        return result;
    }

    @NotNull
    @Override
    public List<JournalRecord> getValuesToMutate(@NotNull List<RecordRef> records) {

        return records.stream()
            .map(RecordRef::getId)
            .map(id -> {
                JournalWithMeta dto = journalService.getJournalById(id);
                if (dto == null) {
                    dto = new JournalWithMeta();
                    dto.setId(id);
                }
                return new JournalRecord(dto);
            })
            .collect(Collectors.toList());
    }

    @NotNull
    @Override
    public RecordsMutResult save(@NotNull List<JournalRecord> values) {

        RecordsMutResult result = new RecordsMutResult();

        for (final JournalRecord model : values) {
            result.addRecord(new RecordMeta(journalService.save(model).getId()));
        }

        return result;
    }

    @Data
    public static class QueryWithTypeRef {
        private String typeRef;
    }

    @Data
    public static class QueryByListId {
        private String listId;
    }

    @Override
    public String getId() {
        return "journal";
    }

    @Data
    public static class JournalQueryByTypeRef {
        private RecordRef typeRef;
    }

    public static class JournalRecord extends JournalWithMeta {

        JournalRecord(JournalWithMeta dto) {
            super(dto);
        }

        public String getModuleId() {
            return getId();
        }

        public void setModuleId(String value) {
            setId(value);
        }

        @Override
        public void setColumns(List<JournalColumnDto> columns) {
            /*
            // todo
            if (columns != null) {
                columns.forEach(c -> {
                    // convert {"key": "{\"json\":\"value\"}"} to {"key": {"json": "value"}}
                    ColumnControl control = c.getControl();
                    if (control != null) {
                        ObjectData config = control.getConfig();
                        if (config.size() > 0) {
                            ObjectData newConfig = ObjectData.create();
                            for (String key : config.fieldNamesList()) {
                                DataValue value = config.get(key);
                                if (value.isTextual()) {
                                    newConfig.set(key, value.asText());
                                } else {
                                    newConfig.set(key, value);
                                }
                            }
                            control.setConfig(newConfig);
                        }
                    }
                });
            }*/
            super.setColumns(columns);
        }

        @Override
        public List<JournalColumnDto> getColumns() {
            return super.getColumns().stream().map(c -> {
                if (c.getControl() == null) {
                    c = new JournalColumnDto(c);
                    ColumnControl columnControl = new ColumnControl();
                    c.setControl(columnControl);
                }
                if (c.getControl().getConfig() == null) {
                    c.getControl().setConfig(ObjectData.create());
                }
                return c;
            }).collect(Collectors.toList());
        }

        @MetaAtt(".type")
        public RecordRef getEcosType() {
            return RecordRef.valueOf("emodel/type@journal");
        }

        @JsonIgnore
        @MetaAtt(".disp")
        public String getDisplayName() {
            String name = MLText.getClosestValue(getLabel(), QueryContext.getCurrent().getLocale());
            return StringUtils.isNotBlank(name) ? name : "Journal";
        }

        @JsonProperty("_content")
        public void setContent(List<ObjectData> content) {

            String base64Content = content.get(0).get("url", "");
            base64Content = base64Content.replaceAll("^data:application/json;base64,", "");
            ObjectData data = Json.getMapper().read(Base64.getDecoder().decode(base64Content), ObjectData.class);

            Json.getMapper().applyData(this, data);
        }

        @JsonValue
        @com.fasterxml.jackson.annotation.JsonValue
        public JournalDto toJson() {
            return new JournalDto(this);
        }
    }
}

File: src/main/java/ru/citeck/ecos/uiserv/domain/journal/api/records/JournalRecordsDao.kt
package ru.citeck.ecos.uiserv.domain.journal.api.records

import com.fasterxml.jackson.annotation.JsonProperty
import com.fasterxml.jackson.annotation.JsonValue
import com.fasterxml.jackson.databind.node.ObjectNode
import jakarta.annotation.PostConstruct
import lombok.Data
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Component
import ru.citeck.ecos.commons.data.DataValue
import ru.citeck.ecos.commons.data.MLText
import ru.citeck.ecos.commons.data.ObjectData
import ru.citeck.ecos.commons.json.Json.mapper
import ru.citeck.ecos.commons.json.YamlUtils
import ru.citeck.ecos.commons.utils.StringUtils
import ru.citeck.ecos.context.lib.auth.AuthContext
import ru.citeck.ecos.context.lib.auth.AuthContext.isRunAsAdmin
import ru.citeck.ecos.events2.type.RecordEventsService
import ru.citeck.ecos.model.lib.utils.ModelUtils
import ru.citeck.ecos.model.lib.workspace.WorkspaceService
import ru.citeck.ecos.records2.RecordConstants
import ru.citeck.ecos.records2.predicate.PredicateService
import ru.citeck.ecos.records3.record.atts.schema.ScalarType
import ru.citeck.ecos.records3.record.atts.schema.annotation.AttName
import ru.citeck.ecos.records3.record.atts.value.AttValue
import ru.citeck.ecos.records3.record.dao.AbstractRecordsDao
import ru.citeck.ecos.records3.record.dao.atts.RecordAttsDao
import ru.citeck.ecos.records3.record.dao.delete.DelStatus
import ru.citeck.ecos.records3.record.dao.delete.RecordDeleteDao
import ru.citeck.ecos.records3.record.dao.mutate.RecordMutateDtoDao
import ru.citeck.ecos.records3.record.dao.query.RecordsQueryDao
import ru.citeck.ecos.records3.record.dao.query.dto.query.RecordsQuery
import ru.citeck.ecos.records3.record.dao.query.dto.res.RecsQueryRes
import ru.citeck.ecos.uiserv.domain.ecostype.service.EcosTypeService
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalActionDef
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalColumnDef
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalDef
import ru.citeck.ecos.uiserv.domain.journal.dto.JournalWithMeta
import ru.citeck.ecos.uiserv.domain.journal.registry.JournalsRegistryConfiguration
import ru.citeck.ecos.uiserv.domain.journal.service.JournalService
import ru.citeck.ecos.uiserv.domain.journal.service.JournalServiceImpl
import ru.citeck.ecos.uiserv.domain.workspace.service.WorkspaceUiService
import ru.citeck.ecos.webapp.api.constants.AppName
import ru.citeck.ecos.webapp.api.entity.EntityRef
import java.nio.charset.StandardCharsets

@Component
@RequiredArgsConstructor
class JournalRecordsDao(
    private val journalService: JournalService,
    private val ecosTypeService: EcosTypeService,
    private val recordEventsService: RecordEventsService,
    private val workspaceService: WorkspaceService,
    private val workspaceUiService: WorkspaceUiService,
) : AbstractRecordsDao(),
    RecordsQueryDao,
    RecordAttsDao,
    RecordMutateDtoDao<JournalRecordsDao.JournalMutateRec>,
    RecordDeleteDao {

    companion object {
        const val ID = "journal"
    }

    @PostConstruct
    fun init() {
        journalService.onJournalChanged { before, after ->
            if (after != null) {
                recordEventsService.emitRecChanged(before, after, getId()) {
                    JournalRecord(JournalWithMeta(it), workspaceService)
                }
            }
        }
    }

    override fun getId() = ID

    override fun queryRecords(recsQuery: RecordsQuery): RecsQueryRes<JournalRecord> {

        if (recsQuery.language == "site-journals") {
            return RecsQueryRes()
        }

        var recsQuery = recsQuery
        val result = RecsQueryRes<JournalWithMeta>()

        if (recsQuery.language == "by-type") {

            val typeRef = recsQuery.getQuery(JournalQueryByTypeRef::class.java).typeRef ?: EntityRef.EMPTY
            val journalRef = ecosTypeService.getJournalRefByTypeRef(typeRef)

            if (EntityRef.isNotEmpty(journalRef)) {
                val idInWs = workspaceService.convertToIdInWs(journalRef.getLocalId())
                val dto = journalService.getJournalById(idInWs)
                if (dto != null) {
                    result.addRecord(JournalRecord(dto, workspaceService))
                }
            }
        } else {

            recsQuery = workspaceUiService.prepareQueryWithSystemFilterIfRequired(recsQuery)

            if (recsQuery.language == PredicateService.LANGUAGE_PREDICATE) {

                val queryRes = recordsService.query(
                    recsQuery.copy()
                        .withSourceId(JournalsRegistryConfiguration.JOURNALS_REGISTRY_SOURCE_ID)
                        .build()
                )

                val journals = journalService.getAll(
                    queryRes.getRecords().map {
                        workspaceService.convertToIdInWs(it.getLocalId())
                    }.toSet()
                )
                result.setRecords(ArrayList(journals))
                result.setTotalCount(queryRes.getTotalCount())
            } else {
                result.setRecords(
                    ArrayList(
                        journalService.getAll(recsQuery.page.maxItems, recsQuery.page.skipCount)
                    )
                )
                result.setTotalCount(journalService.getCount())
            }
        }

        val res = RecsQueryRes<JournalRecord>()
        res.setTotalCount(result.getTotalCount())
        res.setHasMore(result.getHasMore())
        res.setRecords(result.getRecords().map { JournalRecord(it, workspaceService) })
        return res
    }

    override fun getRecordAtts(recordId: String): JournalRecord {
        val dto = if (recordId.isEmpty()) {
            JournalWithMeta(false)
        } else {
            val idInWs = workspaceService.convertToIdInWs(recordId)
            if (!workspaceService.isWorkspaceWithGlobalEntities(idInWs.workspace) &&
                !workspaceService.isUserMemberOf(AuthContext.getCurrentUser(), idInWs.workspace)
            ) {
                JournalWithMeta(false)
            } else {
                journalService.getJournalById(idInWs) ?: JournalWithMeta(false)
            }
        }
        return JournalRecord(dto, workspaceService)
    }

    override fun delete(recordId: String): DelStatus {
        val idInWs = workspaceService.convertToIdInWs(recordId)
        checkWritePermissions(idInWs.workspace)
        journalService.delete(idInWs)
        return DelStatus.OK
    }

    override fun getRecToMutate(recordId: String): JournalMutateRec {
        if (recordId.isEmpty()) {
            return JournalMutateRec(JournalDef.create().build(), workspaceService)
        }
        val idInWs = workspaceService.convertToIdInWs(recordId)
        val dto = journalService.getJournalById(idInWs) ?: error("Record with id '$idInWs' doesn't found")
        return JournalMutateRec(dto.journalDef, workspaceService)
    }

    override fun saveMutatedRec(record: JournalMutateRec): String {
        checkWritePermissions(record.workspace)

        var typeLocalId = record.typeRef.getLocalId()
        if (typeLocalId.isNotBlank()) {
            typeLocalId = workspaceService.replaceCurrentWsPlaceholderToWsPrefix(typeLocalId, record.workspace)
            record.withTypeRef(record.typeRef.withLocalId(typeLocalId))
        }

        record.actionsDef.forEach {
            val config = it.config
            config.forEach { k, v ->
                if (v.isTextual()) {
                    val value = v.asText()
                    if (StringUtils.isNotBlank(value) && value[0] == '{' && value[value.length - 1] == '}') {
                        val convertedValue = DataValue.of(mapper.convert(v, ObjectNode::class.java))
                        if (convertedValue.isObject()) {
                            config[k] = convertedValue
                        }
                    }
                }
            }
        }

        val afterSave = journalService.save(record.build()).journalDef
        return workspaceService.addWsPrefixToId(afterSave.id, afterSave.workspace)
    }

    fun checkWritePermissions(workspace: String?) {
        if (workspaceService.getArtifactsWritePermission(AuthContext.getCurrentUser(), workspace, "journal")) {
            return
        }
        error("Permission denied. You can't create or change journals in workspace '$workspace'")
    }

    @Data
    class JournalQueryByTypeRef(
        val typeRef: EntityRef? = null
    )

    class ActionDefRecord(
        @param:AttName("...")
        val actionDef: JournalActionDef
    ) {
        fun getConfigMap(): Map<String, String> {

            val map = HashMap<String, String>()
            actionDef.config.forEach { k, v ->
                map[k] = if (v.isObject()) {
                    v.toString()
                } else {
                    v.asText()
                }
            }
            return map
        }

        fun getAsJson(): Any {
            return mapper.toNonDefaultJson(actionDef)
        }
    }

    open class JournalRecord(base: JournalWithMeta, val workspaceService: WorkspaceService) : JournalWithMeta(base) {

        @AttName(ScalarType.ID_SCHEMA)
        open fun getRef(): EntityRef {
            val journalId = journalDef?.id ?: ""
            val localId = if (journalId.startsWith("type$")) {
                journalId
            } else {
                workspaceService.addWsPrefixToId(journalId, journalDef?.workspace ?: "")
            }
            return EntityRef.create(AppName.UISERV, ID, localId)
        }

        open fun getModuleId(): String {
            return getLocalId()
        }

        open fun getLocalId(): String {
            return journalDef?.id ?: ""
        }

        @AttName("?type")
        fun getType(): EntityRef = EntityRef.valueOf("emodel/type@journal")

        fun getActionsDef(): List<ActionDefRecord> {
            return journalDef?.actionsDef?.map { ActionDefRecord(it) } ?: emptyList()
        }

        @AttName("?disp")
        fun getDisplayName(): MLText {
            return journalDef?.name ?: MLText.EMPTY
        }

        @JsonValue
        open fun toNonDefaultJson(): Any? {
            val journalDef = journalDef ?: return null
            return mapper.toNonDefaultJson(journalDef.copy().withWorkspace("").build())
        }

        open fun getColumns(): List<Any> {
            return journalDef?.columns?.map { ColumnAttValue(it) } ?: emptyList()
        }

        open fun getData(): ByteArray? {
            val journalDef = journalDef ?: return null
            val journalDefCopy = journalDef.copy {
                withWorkspace("")

                var typeLocalId = journalDef.typeRef.getLocalId()
                if (typeLocalId.isNotBlank()) {
                    typeLocalId = workspaceService.replaceWsPrefixToCurrentWsPlaceholder(typeLocalId)
                    withTypeRef(journalDef.typeRef.withLocalId(typeLocalId))
                }
            }
            return YamlUtils.toNonDefaultString(journalDefCopy).toByteArray(StandardCharsets.UTF_8)
        }

        open fun getPermissions(): Permissions? {
            return Permissions()
        }

        inner class Permissions : AttValue {

            override fun has(name: String): Boolean {
                return if (name.equals("write", ignoreCase = true)) {
                    val journalId: String = getLocalId()
                    if (journalId.contains("$")) {
                        false
                    } else {
                        if (JournalServiceImpl.SYSTEM_JOURNALS.contains(journalId)) {
                            false
                        } else {
                            val workspace = journalDef?.workspace ?: ModelUtils.DEFAULT_WORKSPACE_ID
                            isRunAsAdmin() || workspaceService.isUserManagerOf(AuthContext.getCurrentUser(), workspace)
                        }
                    }
                } else {
                    name.equals("read", ignoreCase = true)
                }
            }
        }
    }

    class ColumnAttValue(
        @param:AttName("...")
        val column: JournalColumnDef
    ) {
        fun getJsonValue(): ObjectData {
            return ObjectData.create(column)
        }
    }

    class JournalMutateRec(
        base: JournalDef,
        private val workspaceService: WorkspaceService
    ) : JournalDef.Builder(base) {

        val originalId = base.id

        fun withModuleId(id: String) {
            withId(id)
        }

        @JsonProperty(RecordConstants.ATT_WORKSPACE)
        fun withCtxWorkspace(workspace: String) {
            if (originalId.isNotBlank() && originalId != id) {
                withWorkspace(workspace)
            } else {
                withWorkspace(workspaceService.getUpdatedWsInMutation(this.workspace, workspace))
            }
        }
    }
}
